### PR TITLE
[Api][Cart] Clear cart

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master 
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 1]
```diff
@@ -46,41 +46,51 @@ public function __construct(
...
-     * @Then my cart should be empty
+     * @Then /^my (cart) should be empty$/
      */
-    public function myCartShouldBeEmpty(): void
+    public function myCartShouldBeEmpty(string $tokenValue): void
     {
-        $response = $this->cartsClient->getLastResponse();
+        $response = $this->cartsClient->show($tokenValue);
+
+        $x = $response;
```
Comment (arti0090) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:79]: Is this $x even used anywhere?

Comment (Tomanhez) [reply src/Sylius/Behat/Context/Api/Shop/CartContext.php:79]: xD

File: src/Sylius/Behat/Client/ResponseChecker.php [hunk 2]
```diff
@@ -55,7 +55,7 @@ public function getError(Response $response): string
 
     public function isCreationSuccessful(Response $response): bool
     {
-        return $response->getStatusCode() === Response::HTTP_CREATED;
+        return $response->getStatusCode() === Response::HTTP_CREATED || $response->getStatusCode() === Response::HTTP_OK;
```
Comment (GSadee) [inline src/Sylius/Behat/Client/ResponseChecker.php:58]: `Response::HTTP_OK` should not be checked in this method, as the method has name `isCreationSuccessful`

File: src/Sylius/Behat/Context/Api/Admin/ManagingAdministratorsContext.php [hunk 3]
```diff
@@ -333,7 +333,7 @@ public function iShouldBeNotifiedThatItCannotBeDeleted(): void
             'Administrator could be deleted'
         );
         Assert::same(
-            $this->responseChecker->getError($this->client->getLastResponse()),
+            $this->responseChecker->getValue($this->client->getLastResponse(), 'message'),
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingAdministratorsContext.php:336]: Why did you change that?

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 4]
```diff
@@ -46,41 +46,50 @@ public function __construct(
...
-        $this->cartsClient->create(Request::create('orders'));
+        $this->cartsClient->show($tokenValue);
     }
 
     /**
-     * @When /^I (?:add|added) (this product) to the cart$/
+     * @When /^I (?:add|added) (this product) to the (cart)$/
      */
-    public function iAddThisProductToTheCart(ProductInterface $product): void
+    public function iAddThisProductToTheCart(ProductInterface $product, $tokenValue): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:59]:
```suggestion
    public function iAddThisProductToTheCart(ProductInterface $product, string $tokenValue): void
```

File: src/Sylius/Behat/Context/Transform/CartContext.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    ) {
+        $this->cartsClient = $cartsClient;
+        $this->responseChecker = $responseChecker;
+        $this->sharedStorage = $sharedStorage;
+    }
+
+    /**
+     * @Transform /^(cart)$/
+     */
+    public function provideCart(string $cart): string
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Transform/CartContext.php:47]: Unnecessary argument

File: src/Sylius/Behat/Context/Transform/CartContext.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        if ($this->sharedStorage->has('cart_token')) {
+            $tokenValue = $this->sharedStorage->get('cart_token');
+
+            $response = $this->cartsClient->show($tokenValue);
+            if ($response->getStatusCode() === Response::HTTP_OK || $response->getStatusCode() === Response::HTTP_CREATED) {
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Transform/CartContext.php:53]:
```suggestion
            if ($response->getStatusCode() === Response::HTTP_OK) {
```
should be enough

File: src/Sylius/Behat/Resources/config/services/contexts/transform.xml [hunk 7]
```diff
@@ -148,6 +148,12 @@
             <argument type="service" id="sylius.behat.shared_storage" />
         </service>
 
+        <service id="sylius.behat.context.transform.cart" class="Sylius\Behat\Context\Transform\CartContext">
+            <argument type="service" id="sylius.behat.api_platform_client.product" />
```
Comment (GSadee) [inline src/Sylius/Behat/Resources/config/services/contexts/transform.xml:152]:
```suggestion
            <argument type="service" id="sylius.behat.api_platform_client.cart" />
```

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderExtension.php [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        if (!is_a($resourceClass, OrderInterface::class, true)) {
+            return;
+        }
+
+        $rootAlias = $queryBuilder->getRootAliases()[0];
+        $user = $this->userContext->getUser();
+
+        $operationName = strtoupper($operationName);
+
+        if ($operationName === Request::METHOD_GET) {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderExtension.php:59]: Extract logic for different operations to private methods

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderExtension.php [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            if ($user === null) {
+                $queryBuilder->andWhere(sprintf('%s.customer IS NULL', $rootAlias));
+
+                return;
+            }
+
+            if ($user instanceof ShopUserInterface && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
+                $queryBuilder
+                    ->andWhere(sprintf('%s.customer = :customerId', $rootAlias))
+                    ->setParameter('customerId', $user->getCustomer()->getId())
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderExtension.php:69]:
```suggestion
                    ->andWhere(sprintf('%s.customer = :customer', $rootAlias))
                    ->setParameter('customer', $user->getCustomer())
```
will be enough

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderExtension.php [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            if ($user instanceof AdminUserInterface && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
+                $queryBuilder
+                    ->andWhere(sprintf('%s.state = :state', $rootAlias))
+                    ->setParameter('state', OrderInterface::STATE_CART)
+                ;
+
+                return;
+            }
+
+            throw new AccessDeniedHttpException(sprintf('Not allowed to %s method', Request::METHOD_DELETE));
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderExtension.php:114]:
```suggestion
            throw new AccessDeniedHttpException('Not allowed to delete method');
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml [hunk 11]
```diff
@@ -40,7 +40,13 @@
 
         <itemOperations>
             <itemOperation name="get">
-                <attribute name="security">is_granted('ROLE_API_ACCESS')</attribute>
+                <attribute name="security">is_granted('IS_AUTHENTICATED_ANONYMOUSLY')</attribute>
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:43]: Is `security` attribute still necessary?

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/PickupCartHandlerSpec.php [hunk 12]
```diff
@@ -35,20 +34,24 @@ final class PickupCartHandlerSpec extends ObjectBehavior
...
+        $this->beConstructedWith($cartFactory, $channelContext, $userContext, $orderManager, $generator);
     }
 
     function it_is_a_message_handler(): void
     {
         $this->shouldImplement(MessageHandlerInterface::class);
     }
 
-    function it_pick_ups_a_cart(
+    function it_pick_ups_a_cart_for_logged_shop_user(
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/PickupCartHandlerSpec.php:49]:
```suggestion
    function it_picks_up_a_cart_for_logged_shop_user(
```

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/PickupCartHandlerSpec.php [hunk 13]
```diff
@@ -62,6 +65,43 @@ function it_pick_ups_a_cart(
...
+        $cart->setCurrencyCode('USD')->shouldBeCalled();
+        $cart->setLocaleCode('en_US')->shouldBeCalled();
+        $cart->setTokenValue('urisafestr')->shouldBeCalled();
+
+        $orderManager->persist($cart)->shouldBeCalled();
+
+        $this(new PickupCart());
+    }
+
+    function it_pick_ups_a_cart_for_visitor(
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/PickupCartHandlerSpec.php:87]:
```suggestion
    function it_picks_up_a_cart_for_visitor(
```

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryItemExtension/OrderExtensionSpec.php [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+
+final class OrderExtensionSpec extends ObjectBehavior
+{
+    function let(UserContextInterface $userContext): void
+    {
+        $this->beConstructedWith($userContext);
+    }
+
+    function it_access_to_get_order_with_null_user_if_present_user_is_null(
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryItemExtension/OrderExtensionSpec.php:34]: IMO there should be sth like `it_applies_condition_...`

File: src/Sylius/Behat/Context/Api/Admin/ManagingAdministratorsContext.php (hunk 3 above)
Comment (Tomanhez) [reply src/Sylius/Behat/Context/Api/Admin/ManagingAdministratorsContext.php:336]: Because its working with collections but not with items

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 15]
```diff
@@ -46,37 +46,59 @@ public function __construct(
...
-    public function myCartShouldBeEmpty(): void
+    public function iAddOfThemToMyCart(int $quantity, ProductInterface $product, string $tokenValue): void
+    {
+        $this->putProductToCart($product, $tokenValue, $quantity);
+    }
+
+    /**
+     * @Then my cart should be cleared
+     */
+    public function myCartShouldBeClean(): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:83]: Not consistent with the step

File: src/Sylius/Behat/Context/Transform/CartContext.php [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        if ($this->sharedStorage->has('cart_token')) {
+            $tokenValue = $this->sharedStorage->get('cart_token');
+
+            $response = $this->cartsClient->show($tokenValue);
+            if ($response->getStatusCode() === Response::HTTP_OK) {
+                return $this->sharedStorage->get('cart_token');
+            }
+        }
+        $response = $this->cartsClient->create(Request::create('orders'));
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Transform/CartContext.php:57]:
```suggestion

        $response = $this->cartsClient->create(Request::create('orders'));
```

File: src/Sylius/Behat/Context/Transform/CartContext.php [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+     * @Transform /^(cart)$/
+     */
+    public function provideCart(): string
+    {
+        if ($this->sharedStorage->has('cart_token')) {
+            $tokenValue = $this->sharedStorage->get('cart_token');
+
+            $response = $this->cartsClient->show($tokenValue);
+            if ($response->getStatusCode() === Response::HTTP_OK) {
+                return $this->sharedStorage->get('cart_token');
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Transform/CartContext.php:54]:
```suggestion
                return $tokenValue;
```
should be enough here

File: src/Sylius/Behat/Context/Transform/CartContext.php [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+    ) {
+        $this->cartsClient = $cartsClient;
+        $this->responseChecker = $responseChecker;
+        $this->sharedStorage = $sharedStorage;
+    }
+
+    /**
+     * @Transform /^(cart)$/
+     */
+    public function provideCart(): string
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Transform/CartContext.php:47]:
```suggestion
    public function provideCartToken(): string
```

File: src/Sylius/Behat/Context/Transform/CartContext.php [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+namespace Sylius\Behat\Context\Transform;
+
+use Behat\Behat\Context\Context;
+use Sylius\Behat\Client\ApiClientInterface;
+use Sylius\Behat\Client\Request;
+use Sylius\Behat\Client\ResponseCheckerInterface;
+use Sylius\Behat\Service\SharedStorageInterface;
+use Symfony\Component\HttpFoundation\Response;
+
+final class CartContext implements Context
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Transform/CartContext.php:23]:
```suggestion
final class CartTokenContext implements Context
```
or sth similar

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderGetMethodExtension.php [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+{
+    /** @var UserContextInterface */
+    private $userContext;
+
+    public function __construct(UserContextInterface $userContext)
+    {
+        $this->userContext = $userContext;
+    }
+
+    public function supports(string $resourceClass, string $operationName = null, array $context = []): bool
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderGetMethodExtension.php:37]: Is this method used here? Because it is not a part of `QueryItemExtensionInterface` if I'm not wrong

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderGetMethodExtension.php [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+        $this->applyToItemForGetMethod($user, $queryBuilder, $operationName, $rootAlias);
+    }
+
+    private function applyToItemForGetMethod(
+        ?UserInterface $user,
+        QueryBuilder $queryBuilder,
+        string $operationName,
+        string $rootAlias
+    ): void {
+        if ($operationName === Request::METHOD_GET) {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderGetMethodExtension.php:68]: Move this `if` to `applyToItem` method next to https://github.com/Sylius/Sylius/pull/11548/files#diff-45ec3630ca31026139f7ff04770f08ccR50 to have all checks in one place
OR
move it to `supports` if it is used

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderDeleteMethodExtension.php [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+    public function applyToItem(
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator,
+        string $resourceClass,
+        array $identifiers,
+        string $operationName = null,
+        array $context = []
+    ) {
+        if (!is_a($resourceClass, OrderInterface::class, true) &&
+            $operationName !== Request::METHOD_DELETE) {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderDeleteMethodExtension.php:46]: I guess that here should be OR instead of AND
```suggestion
        if (
            !is_a($resourceClass, OrderInterface::class, true) ||
            $operationName !== Request::METHOD_DELETE
        ) {
```

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryItemExtension/OrderGetMethodExtensionSpec.php [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+{
+    function let(UserContextInterface $userContext): void
+    {
+        $this->beConstructedWith($userContext);
+    }
+
+    function it_applies_conditions_to_get_order_with_state_cart_and_without_user_if_current_user_is_null(
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator,
+        UserContextInterface $userContext
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryItemExtension/OrderGetMethodExtensionSpec.php:36]: Argument from `let` method should be in the first place (here and in other methods)

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderDeleteMethodExtension.php (hunk 22 above)
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderDeleteMethodExtension.php]: But what happend when `!is_a($resourceClass, OrderInterface::class, true)` return true and `$operationName !== Request::METHOD_DELETE` return false , I think that in this case better would be two `if` statements

For example if `$resourceClass` will be `ShippingMethod::class` and `$operationName` will be `Request::METHOD_GET` , it return `true || false`

**Comment by GSadee** [comment]
Thank you, Tomasz! :tada:
